src/server/scripts/Zandalar/AtalDazar/boss_volkaal.cpp

```diff
@@ -0,0 +1,417 @@
+
+        std::list<Creature*> totems;
+        me->GetCreatureListWithEntryInGrid(totems, NPC_REANIMATION_TOTEM);
+        for (Creature* totem : totems)
+            instance->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, totem);
+
+        Talk(SAY_AGGRO);
+
+        events.ScheduleEvent(EVENT_TOXIC_LEAP, 2200ms);
+        events.ScheduleEvent(EVENT_NOXIOUS_STENCH, 5800ms);
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        _JustDied();
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+
+        Talk(SAY_DIED);
+    }
+
+    void SpellHit(WorldObject* /*caster*/, SpellInfo const* spellInfo) override
+    {
+        if (spellInfo->Id == SPELL_RAPID_DECAY)
+        {
+            Talk(SAY_DECAY);
+            events.CancelEvent(EVENT_TOXIC_LEAP);
+        }
+    }
+
+    void MovementInform(uint32 /*type*/, uint32 id) override
+    {
+        if (id == EVENT_JUMP)
+            DoCastSelf(SPELL_TOXIC_LEAP_DAMAGE);
+    }
+
+    void ExecuteEvent(uint32 eventId) override
+    {
+        switch (eventId)
+        {
+            case EVENT_TOXIC_LEAP:
+            {
+                DoCast(SPELL_TOXIC_LEAP_SELECTOR);
+                events.Repeat(6s);
+                break;
+            }
+            case EVENT_NOXIOUS_STENCH:
+            {
+                DoCast(SPELL_NOXIOUS_STENCH);
+                events.Repeat(19400ms);
+                events.RescheduleEvent(EVENT_TOXIC_LEAP, 8500ms);
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
+
+        events.Update(diff);
+
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
+
+        while (uint32 eventId = events.ExecuteEvent())
+        {
+            ExecuteEvent(eventId);
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+        }
+    }
+};
+
+// 125977 - Reanimation Totem
+struct npc_volkaal_reanimation_totem : public ScriptedAI
+{
+    npc_volkaal_reanimation_totem(Creature* creature) : ScriptedAI(creature) { }
+
+    void JustAppeared() override
+    {
+        if (IsHeroicOrHigher())
+            me->SetUnkillable(true);
+    }
+
+    void DamageTaken(Unit* /*who*/, uint32& damage, DamageEffectType /*damageType*/, SpellInfo const* /*spellInfo = nullptr*/) override
+    {
+        if (IsHeroicOrHigher())
+        {
+            if (me->GetHealth() <= damage)
+            {
+                DoCastSelf(SPELL_SOUL_ANCHOR);
+                DoCastSelf(SPELL_REANIMATE);
+            }
+        }
+
+        InstanceScript* instance = me->GetInstanceScript();
+        if (!instance)
+            return;
+
+        if (Creature* volkaal = instance->GetCreature(DATA_VOLKAAL))
+            volkaal->AI()->DoZoneInCombat();
+    }
+
+    void DoAction(int32 action) override
+    {
+        if (action == ACTION_TOTEMS_DIED)
```

Comment:
with volkaal having the totems in summons we could move this action to boss script itself, otherwise this action is rather unnecessary and can be executed without DoAction wherever its needed